mantidproject/mantid
# Fixes for reflectometry workflow algorithms history

**Report to:** Jos

This PR fixes some issues with workspace history for interim output workspaces from the ISIS Reflectometry workflow algorithms. It includes:
- Addition of output properties for the interim workspaces to replace direct use of the ADS.
- Improved grouping of TOF workspaces. Previously only the input TOF workspaces were grouped, but now monitors are also grouped.
- A bug fix where the transmission workspace was being dumped before it was rebinned in `ReflectometryReductionOne::transmissionCorrection`, so there were minor binning differences in the workspace seen in the ADS to the one that was used in the correction.
- Minor renaming and tidying.

**To test:**
Set the default instrument to `INTER`. For each case in the table below:
- Run the code in the table to set up the arguments, and then use them to run the algorithm like this:
  ```
  ReflectometryISISLoadAndProcess(**args)
  ```
- Check the history of `IvsQ_binned_13460`. The following mandatory outputs should be set for every case:
  - OutputWorkspaceBinned=`IvsQ_binned_13460`
  - OutputWorkspace=`IvsQ_13460`
- The additional optional outputs that should also be set for each case are listed in the table. You can use this script will print the output properties:
  ```
  properties=['OutputWorkspaceBinned', 'OutputWorkspace', 'OutputWorkspaceWavelength',, 'OutputWorkspaceTransmission', 'OutputWorkspaceFirstTransmission', 'OutputWorkspaceSecondTransmission']
  for property in properties:
      print(property + ' = ' + mtd['IvsQ_binned_13460'].getHistory().getAlgorithmHistory(0).getPropertyValue(property))
  ```
- Where a `TRANS_LAM...` workspace is output, check that it has history. Note that the loaded runs in TOF do not have history.

Script | Optional outputs
-------|-----------------
`args={'Debug':0, 'InputRunList':'13460'}` | None
`args={'Debug':1, 'InputRunList':'13460'}` | OutputWorkspaceWavelength=`IvsLam_13460`
`args={'Debug':0, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463'}` | OutputWorkspaceTransmission=`TRANS_LAM_13463`
`args={'Debug':1, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463'}` | OutputWorkspaceWavelength=`IvsLam_13460` OutputWorkspaceTransmission=`TRANS_LAM_13463`
`args={'Debug':0, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463', 'SecondTransmissionRunList':'13464'}` | OutputWorkspaceTransmission=`TRANS_LAM_13463_13464`
`args={'Debug':1, 'InputRunList':'13460', 'FirstTransmissionRunList':'13463', 'SecondTransmissionRunList':'13464'}`| OutputWorkspaceWavelength=`IvsLam_13460` OutputWorkspaceTransmission=`TRANS_LAM_13463_13464` OutputWorkspaceFirstTransmission=`TRANS_LAM_13463` OutputWorkspaceSecondTransmission=`TRANS_LAM_13464`

Fixes #27633 

Release notes to follow

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
